### ISSUE TITLE
The mean method results in different outputs

### ISSUE THREAD

**Issue by KaiHoo** [pytorch/pytorch#36582]
## 🐛 Bug

The results of mean method of torch.Tensor instance would be wrong when the tensor is big.

## To Reproduce

```
import torch

b = torch.randn(1024,224,224,3)
b = b.sub_(b.min())

b.mean((0,1,2))[0]
b[:,:,:,0].mean()
b.permute(0,3,1,2).contiguous().mean((0,2,3))[0]
b.permute(0,3,1,2).mean((0,2,3))[0]

# if cuda is available
b = b.cuda()
b.mean((0,1,2))[0]
b[:,:,:,0].mean()
b.permute(0,3,1,2).contiguous().mean((0,2,3))[0]
b.permute(0,3,1,2).mean((0,2,3))[0]
```
## Expected behavior
If torch is compiled without GPU, the results of ```b.mean((0,1,2))[0]```, ```b[:,:,:,0].mean()``` and ```b.permute(0,3,1,2).mean((0,2,3))[0]``` are the same, but they are different from the result of ```b.permute(0,3,1,2).contiguous().mean((0,2,3))[0]```. I think only the result of ```b.permute(0,3,1,2).contiguous().mean((0,2,3))[0]``` is correct.

If torch is compiled with GPU, the results of ```b.mean((0,1,2))[0]``` and ```b.permute(0,3,1,2).mean((0,2,3))[0]``` are the same, which are not correct. The results of ```b[:,:,:,0].mean()``` and ```b.permute(0,3,1,2).contiguous().mean((0,2,3))[0]```   are the same.

If the tensor is moved to GPU, all four results are the same and correct.

## Environment

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 9.2

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.4.0
[pip3] torchvision==0.5.0
[conda] blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] cudatoolkit               9.2                           0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl                       2020.0                      166    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-service               2.3.0            py37he904b0f_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.0.15           py37ha843d7b_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_random                1.1.0            py37hd6b4f25_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy                     1.18.1                   pypi_0    pypi
[conda] numpy-base                1.18.1           py37hde5b4d6_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] pytorch                   1.4.0           py3.7_cuda9.2.148_cudnn7.6.3_0    pytorch
[conda] torchvision               0.5.0                 py37_cu92    pytorch

## Additional context

The bug exists from torch==1.1.0 to torch==1.4.0, and seems to happen only on CPU.

**Comment by colesbury** [comment]
Your tensor is too large to accurately compute the mean or sum using float32. I notice that NumPy also gives inaccurate results for many of the same computations:

```python
a = b.numpy()
a.mean((0,1,2))[0]  # inaccurate
a[:,:,:,0].mean()      # accurate
np.ascontiguousarray(a.transpose(0,3,1,2)).mean((0,2,3))[0] # accurate
a.transpose(0,3,1,2).mean((0,2,3))[0] # inaccurate
```

Use double precision if you need to compute the mean over large tensors, e.g.:

```python
b.mean((0,1,2), dtype=torch.double)[0]
```

The GPU has a high degree of parallelism, which leads to a lot of partial pairwise summation and consequentially more accurate results.

**Comment by KaiHoo** [comment]
Thanks for your response! I found that even a tensor of shape (256, 224, 224, 3) would lead to 1.6% relative error. This shape should be common for vision tasks, and the mean operation is also common (for batchnorm). It would be helpful if this problem could be solved.

BTW, why does contiguous method help in this case?

**Comment by colesbury** [comment]
> BTW, why does contiguous method help in this case?

I haven't looked closely, but my guess is that it forces multiple smaller sums. The 4th and 1st means get implemented as the same thing basically b.reshape(1024*224*224, 3).sum(0)/N. The 3rd gets implemented as tmp.sum(0).sum((2,3)) / N.

If you need batch normalization, use the built-in batch normalization function. Otherwise, use double precision as I suggested above. I don't think this is going to be improved anytime soon.
